gateway/mw_graphql_granular_access_test.go

```diff
@@ -0,0 +1,89 @@
+package gateway
+
+import (
+	"net/http"
+	"testing"
+
+	"github.com/jensneuse/graphql-go-tools/pkg/graphql"
+
+	"github.com/TykTechnologies/tyk/headers"
+	"github.com/TykTechnologies/tyk/test"
+	"github.com/TykTechnologies/tyk/user"
+)
+
+func TestGraphQL_RestrictedTypes(t *testing.T) {
+	g := StartTest()
+	defer g.Close()
+
+	api := BuildAndLoadAPI(func(spec *APISpec) {
+		spec.Proxy.ListenPath = "/"
+		spec.UseKeylessAccess = false
+		spec.GraphQL.Enabled = true
+	})[0]
+
+	_, directKey := g.CreateSession(func(s *user.SessionState) {
+		s.AccessRights = map[string]user.AccessDefinition{
+			api.APIID: {
+				APIID:   api.APIID,
+				APIName: api.Name,
+				RestrictedTypes: []graphql.Type{
+					{
+						Name:   "Country",
+						Fields: []string{"code"},
+					},
+				},
+			},
+		}
+	})
+
+	pID := CreatePolicy(func(p *user.Policy) {
+		p.AccessRights = map[string]user.AccessDefinition{
+			api.APIID: {
+				APIID:   api.APIID,
+				APIName: api.Name,
+				RestrictedTypes: []graphql.Type{
+					{
+						Name:   "Country",
+						Fields: []string{"name"},
+					},
+				},
+			},
+		}
+	})
+
+	_, policyAppliedKey := g.CreateSession(func(s *user.SessionState) {
+		s.ApplyPolicies = []string{pID}
+	})
+
+	q1 := graphql.Request{
```

Comment:
maybe better `unrestrictedQuery` and `restrictedQuery` ?

Reply:
agree!